star-sys/STAR/source/stitchAlignToTranscript.cpp

```diff
@@ -121,28 +123,49 @@ intScore stitchAlignToTranscript(uint rAend, uint gAend, uint rBstart, uint gBst
                     int Score2=Score1;
 
                     if (Del>=P.alignIntronMin) {//only check intron motif for large gaps= non-Dels
-                        //check if the intron is canonical, or semi-canonical
-                        if ( G[gAend+jR1+1]==2 && G[gAend+jR1+2]==3 && G[gBstart1+jR1-1]==0 && G[gBstart1+jR1]==2 ) {//GTAG
-                            jCan1=1;
-                        } else if ( G[gAend+jR1+1]==1 && G[gAend+jR1+2]==3 && G[gBstart1+jR1-1]==0 && G[gBstart1+jR1]==1 ) {//CTAC
-                            jCan1=2;
-                        } else if ( G[gAend+jR1+1]==2 && G[gAend+jR1+2]==1 && G[gBstart1+jR1-1]==0 && G[gBstart1+jR1]==2 ) {//GCAG
-                            jCan1=3;
-                            jPen1=P.scoreGapGCAG;
-                        } else if ( G[gAend+jR1+1]==1 && G[gAend+jR1+2]==3 && G[gBstart1+jR1-1]==2 && G[gBstart1+jR1]==1 ) {//CTGC
+
+                        // Somewhat convoluted logic below to optimize this code path
+                        // in order to check if the intron is canonical, or semi-canonical
+                        // Basically, we're looking for specific splice junctions 
+                        // and manually doing some optimizations it seems the compiler failed at.
+
+                        char first_bp = G[gAend+jR1+1];
+                        char second_bp = G[gAend+jR1+2];
+                        char third_bp = G[gBstart1+jR1-1];
+                        char fourth_bp = G[gBstart1+jR1];
+
+                        // Check 3rd basepair first
+                        if (third_bp==0) {    // NNAN
```

Comment:
The profiler showed that this section of code was being called `A LOT`, I suspect that the right answer here is just not to call this method in the inner loop which I suspect we're doing too often for bad reads, but for the moment just optimized it as is.

Reply:
I'm not actually convinced what you're doing here will actually speed it up.  Branching is quite expensive.  In fact I suspect the most expensive part of what it was doing before was using `&&` rather than `&`.  In particular, separating the expressions like this means the compiler can't easily fuse comparisons of two `char`s into a single `int16_t` comparison, which it was probably doing before.